### PR TITLE
Multilevel caches

### PR DESCRIPTION
**Pull request by rsdy**
This PR improves on our index cache situation by making it symbolic, use an SQL backend, and granular enough to track itra-file differences, speeding up the indexing process for small changesets or metadata changes.

List of changes:

 * Add new branch-aware file-level cache
 * Recognize and update existing embeddings in qdrant
 * Add branch selector to UI with ability to sync new branches
 * Status reporting for incremental & branch indexing
 * Make all the querying API surface aware of branches
 * Ensure consistent query behaviour for filters specified in an LLM conversation.

### CODE REVIEW

File: server/bleep/migrations/20230613143506_file-cache.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- Add migration script here
+CREATE TABLE file_cache (
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    repo_ref TEXT NOT NULL,
+    file_name TEXT NOT NULL,
+    hash TEXT NOT NULL,
```
Comment (ggordonhall) [inline server/bleep/migrations/20230613143506_file-cache.sql:6]: Hash of what?

File: server/bleep/src/db.rs [hunk 2]
```diff
@@ -9,8 +9,7 @@ use crate::Configuration;
 pub type SqlDb = Arc<SqlitePool>;
 
 pub async fn init(config: &Configuration) -> Result<SqlitePool> {
-    fs::create_dir_all(&config.data_dir)?;
-    let data_dir = config.data_dir.to_string_lossy();
+    let data_dir = config.index_dir.to_string_lossy();
```
Comment (ggordonhall) [inline server/bleep/src/db.rs:12]: Why remove `create_dir_all`?

File: server/bleep/src/indexes.rs [hunk 3]
```diff
@@ -66,18 +68,14 @@ impl<'a> GlobalWriteHandle<'a> {
...
 
-        tokio::task::block_in_place(|| {
-            self.handles
-                .par_iter()
-                .map(|handle| {
-                    handle.index(&sync_handle.reporef, repo, &metadata, sync_handle.pipes())
-                })
-                .collect::<Result<Vec<_>, _>>()
-        })?;
+        futures::future::join_all(self.handles.iter().map(|handle| {
```
Comment (ggordonhall) [inline server/bleep/src/indexes.rs:73]: Why the change here?

File: server/bleep/src/indexes/file.rs [hunk 4]
```diff
@@ -637,6 +627,36 @@ impl RepoFile {
...
+        }
+        Entry::Occupied(mut val) => {
+            _ = val.insert((content_hash, branch_list.to_owned()).into());
+        }
+        Entry::Vacant(val) => {
+            _ = val.insert_entry((content_hash, branch_list.to_owned()).into());
+        }
+    }
+
+    trace!("cache miss");
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:656]: When would this happen?

File: server/bleep/src/semantic.rs [hunk 5]
```diff
@@ -65,29 +67,38 @@ macro_rules! val_parse_str(($hash:ident, $val:expr) => {
...
             ..
         } = orig;
 
-        let mut converted = payload
-            .into_iter()
-            .map(|(key, value)| (key, kind_to_value(value.kind)))
-            .collect::<HashMap<String, serde_json::Value>>();
+        let Some(PointId { point_id_options: Some(PointIdOptions::Uuid(id)) }) = id
+	else {
+	    // unless the db was corrupted/written by someone else,
```
Comment (ggordonhall) [inline server/bleep/src/semantic.rs:82]: By `db` are we referring to `Qdrant` here?

File: server/bleep/src/cache.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+impl<'a> ChunkCache<'a> {
+    pub async fn for_file(
+        qdrant: &'a QdrantClient,
+        repo_ref: &'a str,
+        repo_name: &'a str,
+        relative_path: &'a str,
+    ) -> anyhow::Result<ChunkCache<'a>> {
+        let response = qdrant
+            .search_points(&SearchPoints {
```
Comment (ggordonhall) [inline server/bleep/src/cache.rs:152]: Can we search Qdrant by `PointId` rather than by vector here? Where the `id` is a hash of identifying fields.

Comment (rsdy) [reply server/bleep/src/cache.rs:152]: No, because we don't know the identifying fields for chunks that are not currently being indexed. It's a null-vector that we're using for search, because this is the only way we can filter for attributes.

File: server/bleep/src/indexes/file.rs (hunk 4 above)
Comment (rsdy) [reply server/bleep/src/indexes/file.rs:656]: When the item isn't in the cache.

File: server/bleep/src/db.rs (hunk 2 above)
Comment (rsdy) [reply server/bleep/src/db.rs:12]: Since `data_dir` is redundant, we can rely on `index_dir` existing at the time this is created.

File: server/bleep/src/cache.rs (hunk 6 above)
Comment (nerdypepper) [reply server/bleep/src/cache.rs:152]: if i understand, this search operation on qdrant produces all the points for a given file? can we use [`QdrantClient::scroll`](https://docs.rs/qdrant-client/latest/qdrant_client/client/struct.QdrantClient.html#method.scroll) instead?

File: server/bleep/src/cache.rs [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        &self,
+        data: &'a str,
+        embedder: impl FnOnce(&'a str) -> anyhow::Result<Embedding>,
+        mut payload: Payload,
+    ) -> anyhow::Result<()> {
+        let id = {
+            let mut bytes = [0; 16];
+            bytes.copy_from_slice(&blake3::hash(data.as_ref()).as_bytes()[16..32]);
+            Uuid::from_bytes(bytes).to_string()
+        };
```
Comment (nerdypepper) [inline server/bleep/src/cache.rs:202]: does this reduce a 32-byte hash to a 16-byte uuid? how resistant is this to collisions?

**Review by nerdypepper — CHANGES_REQUESTED** [review]

File: server/bleep/src/indexes.rs (hunk 3 above)
Comment (rsdy) [reply server/bleep/src/indexes.rs:73]: It turned into async.

File: server/bleep/src/semantic.rs (hunk 5 above)
Comment (rsdy) [reply server/bleep/src/semantic.rs:82]: Yes.

File: server/bleep/src/cache.rs [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+            .execute(&mut tx)
+            .await?;
+        }
+
+        tx.commit().await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn delete_for_repo(&self, reporef: &RepoRef) -> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (rsdy) [reply server/bleep/src/cache.rs:113]: The naming may be misleading, `delete_all` works on a `Transaction` and it's an implementation detail, `delete_for_repo` is public and creates a new transaction.

File: server/bleep/src/cache.rs (hunk 7 above)
Comment (rsdy) [reply server/bleep/src/cache.rs:202]: Since `blake3` is a cryptographic hash, it should be pretty resistant to collisions. Reducing the 256bit output to 128 bit would still require more than 2^64 points in the db for a 50% chance of collision (according to the birthday paradox), so I don't think this is something we need to worry about too much.

File: server/bleep/src/indexes/file.rs [hunk 9]
```diff
@@ -399,28 +390,21 @@ impl File {
...
-                _ = val.insert(content_hash.into());
-            }
-            Entry::Vacant(val) => {
-                _ = val.insert_entry(content_hash.into());
-            }
+        if matches!(dir_entry, RepoDirEntry::File(_))
+            && is_cache_fresh(cache, &unique_hash, &entry_pathbuf, &dir_entry)
+        {
+            info!("fresh; skipping");
+            return Ok(());
```
[the remark this answers is not in the document]
Comment (rsdy) [reply server/bleep/src/indexes/file.rs]: Good spot, thanks.

**Review by nerdypepper — APPROVED** [review]
lgtm. will like to test this a bit.

File: server/bleep/src/cache.rs [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    }
+}
+
+impl<T> From<T> for FreshValue<T> {
+    fn from(value: T) -> Self {
+        Self { fresh: true, value }
+    }
+}
+
+pub(crate) type RepoCacheSnapshot = Arc<scc::HashMap<String, FreshValue<()>>>;
```
Comment (ggordonhall) [inline server/bleep/src/cache.rs:52]: Can you add comments explaining what these different levels of caching do?
- `RepoCacheSnapshot`
- `FileCache`
- `ChunkCache`

File: server/bleep/src/indexes/file.rs [hunk 11]
```diff
@@ -44,26 +44,27 @@ struct Workload<'a> {
...
         &self,
         reporef: &RepoRef,
         repo: &Repository,
         repo_metadata: &RepoMetadata,
         writer: &IndexWriter,
         pipes: &SyncPipes,
     ) -> Result<()> {
-        let file_cache = repo.open_file_cache(&self.config.index_dir);
+        let file_cache = FileCache::new(&self.sql);
+        let repo_cache = file_cache.for_repo(reporef).await?;
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:62]: This is confusingly named as it's still a file cache. Can you rename to `file_cache_by_repo`

File: server/bleep/src/indexes/file.rs [hunk 12]
```diff
@@ -194,17 +171,33 @@ impl Indexer<File> {
         // hits is a mapping between a document address and the number of trigrams in it that
         // matched the query
         let repo_ref_term = Term::from_field_text(self.source.repo_ref, &repo_ref.to_string());
+        let branch_term = branch
+            .map(|b| {
+                trigrams(b)
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:176]: Would this match two branches which have the same set of trigrams in a different order?

File: server/bleep/src/indexes/file.rs [hunk 13]
```diff
@@ -285,20 +279,26 @@ impl Indexer<File> {
         // XXX: can we use the bloop query language here instead?
         let query_parser = QueryParser::for_index(
             file_index,
-            vec![self.source.repo_disk_path, self.source.relative_path],
+            vec![self.source.repo_ref, self.source.relative_path],
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:282]: Why this change?

File: server/bleep/src/indexes/file.rs [hunk 14]
```diff
@@ -338,17 +338,34 @@ impl Indexer<File> {
...
-        let repo_query = Box::new(TermQuery::new(
+        query.push(Box::new(TermQuery::new(
             Term::from_field_text(self.source.repo_ref, &repo_ref.to_string()),
             IndexRecordOption::Basic,
-        ));
+        )) as Box<dyn Query>);
+
+        let branch_term = branch
+            .map(|b| {
+                trigrams(b)
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:356]: Again, would this match two branches with the same set of trigrams?

File: server/bleep/src/indexes/file.rs [hunk 15]
```diff
@@ -408,42 +423,40 @@ impl File {
         };
         let entry_pathbuf = repo_disk_path.join(&relative_path);
 
-        let content_hash = {
+        let semantic_hash = {
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:426]: Do we need this `semantic_hash` if we select by `repo_name` and `relative_path` in Qdrant?

File: server/bleep/src/indexes/file.rs [hunk 16]
```diff
@@ -523,13 +538,13 @@ impl RepoDir {
...
                 schema.raw_relative_path => relative_path_str.as_bytes(),
                 schema.repo_disk_path => repo_disk_path.to_string_lossy().as_ref(),
-                schema.entry_disk_path => entry_pathbuf.to_string_lossy().as_ref(),
                 schema.relative_path => relative_path_str,
                 schema.repo_ref => repo_ref,
                 schema.repo_name => repo_name,
                 schema.last_commit_unix_seconds => last_commit,
                 schema.branches => branches,
                 schema.is_directory => true,
+                schema.unique_hash => tantivy_cache_key,
```
Comment (ggordonhall) [inline server/bleep/src/indexes/file.rs:547]: Can you give this a more descriptive name?

File: server/bleep/src/semantic.rs [hunk 17]
```diff
@@ -395,19 +430,21 @@ impl Semantic {
         Ok(deduplicate_snippets(results, vector, limit))
     }
 
+    #[allow(clippy::too_many_arguments)]
     #[tracing::instrument(skip(self, repo_ref, relative_path, buffer))]
     pub async fn insert_points_for_buffer(
         &self,
         repo_name: &str,
         repo_ref: &str,
+        unique_hash: &str,
```
Comment (ggordonhall) [inline server/bleep/src/semantic.rs:439]: Not clear what this is a hash of

File: server/bleep/src/semantic.rs [hunk 18]
```diff
@@ -420,73 +457,66 @@ impl Semantic {
...
-                        .into_qdrant(),
-                    }),
-                    Err(err) => {
-                        warn!(?err, %chunk_prefix, "embedding failed");
-                        None
-                    }
-                },
-            )
-            .collect::<Vec<_>>();
+        let embedder = |c: &str| {
```
Comment (ggordonhall) [inline server/bleep/src/semantic.rs:460]: Do we need this?

File: server/bleep/src/semantic.rs [hunk 19]
```diff
@@ -420,73 +457,66 @@ impl Semantic {
...
+            info!("generating embedding");
+            self.embed(c)
+        };
+        chunks.par_iter().for_each(|chunk| {
+            let data = format!(
+                "{repo_name}\t{relative_path}\n{unique_hash}\n{}",
```
Comment (ggordonhall) [inline server/bleep/src/semantic.rs:466]: Why is `unique_hash` here? This could negatively impact retrieval.

File: server/bleep/src/webserver/answer/context.rs [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+        debug!(?query, %self.thread_id, "executing semantic query");
+        self.app
+            .semantic
+            .as_ref()
+            .unwrap()
+            .search(&query, limit, offset, retrieve_more)
+            .await
+    }
+
+    pub(super) async fn file_search(&self, path: &str) -> Result<ContentDocument> {
```
Comment (ggordonhall) [inline server/bleep/src/webserver/answer/context.rs:127]:
```suggestion
    pub(super) async fn get_file_contents(&self, path: &str) -> Result<ContentDocument> {
```

**Review by ggordonhall — CHANGES_REQUESTED** [review]

File: server/bleep/src/background.rs [hunk 21]
```diff
@@ -223,7 +286,7 @@ impl BoundSyncQueue {
         self,
         reporef: RepoRef,
     ) -> anyhow::Result<SyncStatus> {
-        let handle = SyncHandle::new(self.0.clone(), reporef, self.1.progress.clone());
+        let handle = SyncHandle::new(self.0.clone(), reporef, self.1.progress.clone(), None);
```
[the remark this answers is not in the document]
Comment (rsdy) [reply server/bleep/src/background.rs:289]: It's not a single-branch alternative, it's just re-indexing with the existing filters (ie, not changing anything other than Git pulling the repo). We track the difference of branch filters in a `SyncHandle` to help report accurate status to the frontend through the various endpoints that are exposed.

File: server/bleep/src/cache.rs [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+    qdrant: &'a QdrantClient,
+    cache: scc::HashMap<String, FreshValue<Payload>>,
+    update: RwLock<Vec<(PointsSelector, qdrant_client::client::Payload)>>,
+    new: RwLock<Vec<PointStruct>>,
+}
+
+impl<'a> ChunkCache<'a> {
+    pub async fn for_file(
+        qdrant: &'a QdrantClient,
+        content_hash: &'a str,
```
[the remark this answers is not in the document]
Comment (rsdy) [reply server/bleep/src/cache.rs:149]: `repo_ref` and `relative_path` are not sufficient to track unique files in a git repository when we're aware of branches. There are 2 kinds of deterministic unique ids we use: one for tantivy, which is file-level, and one in qdrant which is chunk-level. How those IDs are composed/created is not particularly interesting or relevant at this layer, only that they're a unique reference to a file based in a content-addressed storage space.

File: server/bleep/src/cache.rs [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+    ) -> anyhow::Result<ChunkCache<'a>> {
+        let response = qdrant
```
[the remark this answers is not in the document]
Comment (rsdy) [reply server/bleep/src/cache.rs:151]: We may be able to further limit the fields in the payload we retrieve to just the relevant fields. Upon another glance at the docs, their `set_payload` and `overwrite_payload` functions are separate, so we don't need to send the full payload back, and can selectively patch the different fields.

File: server/bleep/src/indexes/file.rs (hunk 13 above)
Comment (rsdy) [reply server/bleep/src/indexes/file.rs:282]: Moves our logic from disk-based to a purely symbolic space, consistent with removing IDs that carry semantic meaning.

File: server/bleep/src/indexes/file.rs (hunk 15 above)
Comment (rsdy) [reply server/bleep/src/indexes/file.rs:426]: As noted above, the combination of `repo_name` and `relative_path` is insufficient to uniquely identify resources when dealing with branches.

File: server/bleep/src/indexes/file.rs (hunk 16 above)
Comment (rsdy) [reply server/bleep/src/indexes/file.rs:547]: It's not intended to carry semantic meaning. It needs to fulfill 2 criteria:

 * db-wide unique
 * deterministic
 
Relying on or expressing any other incidental property these IDs may carry is not a good idea.

File: server/bleep/src/semantic.rs (hunk 18 above)
Comment (rsdy) [reply server/bleep/src/semantic.rs:460]: We lazily generate embeddings, and we need to pass *a* lambda or object ref either way. Now the main purpose is to allow logging from this callback with all the associated context in the logging span.

File: server/bleep/src/semantic.rs (hunk 19 above)
Comment (rsdy) [reply server/bleep/src/semantic.rs:466]: It's being used to pin the chunk to the containing file. I can't reason either way about retrieval performance (although I think this is in the *probably fine* territory), but I guess we could construct the reference hash differently, and achieve the same effect.

Comment (ggordonhall) [reply server/bleep/src/semantic.rs:466]: We don't want to embed it. `repo_name` and `relative_path` are included in `data` to improve retrieval (by matching on these fields).

File: server/bleep/src/cache.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (rsdy) [reply server/bleep/src/cache.rs:152]: The other nuance here is that this algo will not really impact cold indexing times, as the caches will be empty. The only overhead we rack up here is around calling qdrant, which we can't avoid without a dedicated fastpath if we somehow propagate that this is a cold index.

File: server/bleep/src/indexes/file.rs (hunk 12 above)
Comment (rsdy) [reply server/bleep/src/indexes/file.rs:176]: I've tried resolving this issue, but moving to a [whitespace tokenizer](https://github.com/BloopAI/bloop/commit/36722257b49570f0588135f5eabc4c6fca467879) triggers an infinite loop in tantivy that eats up all the memory. We've seen this bug in passing, but here it's reliable.

Will stick with this approach for now, and revisit this once we upgrade tantivy to 0.20+

Comment (nerdypepper) [reply server/bleep/src/indexes/file.rs:176]: >Would this match two branches which have the same set of trigrams in a different order?

yes, fixable by turning branches to tokenize differently, or by applying a regex filter after the trigram matching work.

File: server/bleep/src/webserver/answer.rs [hunk 24]
```diff
@@ -596,6 +616,9 @@ struct Agent {
     thread_id: uuid::Uuid,
     query_id: uuid::Uuid,
 
+    query: String,
+    branch: Option<String>,
```
Comment (calyptobai) [inline server/bleep/src/webserver/answer.rs:620]: Can we move `branch` to `Conversation`? If a single conversation can span multiple branches, switching between exchanges, then can we move it to `Exchange` instead?

File: server/bleep/src/webserver/answer.rs [hunk 25]
```diff
@@ -596,6 +616,9 @@ struct Agent {
     thread_id: uuid::Uuid,
     query_id: uuid::Uuid,
 
+    query: String,
```
Comment (calyptobai) [inline server/bleep/src/webserver/answer.rs:619]: Do we actually need to record this at all? We have the `branch` already, can't we just add that in `semantic_query_params`?

Or, if we do want to include the other fields like language and path, can we store those directly like `branch`?

**Review by calyptobai — CHANGES_REQUESTED** [review]
I've merged in `main`, and moved some data around. Mostly it is the same, several values from `AppContext` were moved to `Agent`. Have left a few comments about the data added.

File: server/bleep/src/webserver/answer.rs (hunk 24 above)
Comment (rsdy) [reply server/bleep/src/webserver/answer.rs:620]: @ggordonhall ?

File: server/bleep/src/webserver/answer.rs (hunk 25 above)
Comment (rsdy) [reply server/bleep/src/webserver/answer.rs:619]: The idea would have been to record _all_ filters we recognize in the ql, and pass them on. The raw string is stored because `SemanticQuery` is referential, and we can't clone it to owned. `branch` is a separate field so that we can pass it to tantivy queries.

It would be great if we could store an owned `SemanticQuery` or the sort so we don't need to break out the individual values (and re-construct them manually when anything changes), but I didn't have a good idea how to do that.

File: server/bleep/src/webserver/answer.rs (hunk 24 above)
Comment (ggordonhall) [reply server/bleep/src/webserver/answer.rs:620]: A `Conversation` can span and switch between multiple branches, so we should move it to `Exchange`

File: server/bleep/src/webserver/answer.rs (hunk 25 above)
Comment (calyptobai) [reply server/bleep/src/webserver/answer.rs:620]: In that case, can we store an owned `SemanticQuery` on the `Exchange`, in order to capture these other filters? As noted [above](https://github.com/BloopAI/bloop/pull/639#discussion_r1254099183), we know that `branch` is per-exchange. Rather than storing each filter separately on `Exchange`, this would entail storing an owned `SemanticQuery<'static>` (which encapsulates the `branch` set).

We can create an owned value of type `SemanticQuery<'static>` with a method like so:

```rust
impl SemanticQuery<'a> {
    pub fn into_owned(self) -> SemanticQuery<'static> {
        // Recursively create this method on all lifetime bearers, terminating by creating an
        // owned `Cow<'static, str>` in `Literal`.

        // ...
    }
}
```

And we can create `Cow<'static, str>` from `Cow<'a, str>` value like so:

```rust
fn example<'a>(cow: Cow<'a, str>) -> Cow<'static, str> {
    Cow::Owned(cow.into_owned())
}
```

File: server/bleep/src/ee.rs [hunk 26]
```diff
@@ -0,0 +1 @@
+pub(crate) mod webserver;
```
Comment (calyptobai) [inline server/bleep/src/ee.rs:1]: Can we rename the `ee` module to something less opaque?

Comment (ggordonhall) [reply server/bleep/src/ee.rs:1]: `ee` is a bit of an industry standard. See: 
- https://github.com/windmill-labs/windmill/blob/main/backend/src/ee.rs
- https://github.com/strapi/strapi/tree/main/packages/core/admin/ee

File: server/bleep/src/webserver/answer.rs (hunk 24 above)
Comment (ggordonhall) [reply server/bleep/src/webserver/answer.rs]: @calyptobai Have pushed this, can you check it out?

File: server/bleep/src/webserver/answer.rs (hunk 25 above)
Comment (ggordonhall) [reply server/bleep/src/webserver/answer.rs:620]: @calyptobai If this is a quick change can you take this on now?

File: server/bleep/src/ee/webserver.rs [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+
+#[derive(Deserialize)]
+pub(crate) struct RepositoryPatch {
+    branch_filter: BranchFilter,
+}
+
+/// Patch a repository with the given payload
+/// This will automatically trigger a sync
+//
+pub(crate) async fn patch_indexed(
```
Comment (ggordonhall) [inline server/bleep/src/ee/webserver.rs:24]: Could this be used for deleting as well as syncing branches? We should rename to something branch specific like `sync_branch` or `patch_branch` (if deletes might be supported).

**Review by ggordonhall — DISMISSED** [review]

File: server/bleep/src/webserver/answer.rs (hunk 25 above)
Comment (calyptobai) [reply server/bleep/src/webserver/answer.rs]: Resolved in [3d772c3](https://github.com/BloopAI/bloop/pull/639/commits/3d772c307205938ca18ea42c5ad0f34ed4ce7490)

File: server/bleep/src/webserver/answer.rs (hunk 24 above)
Comment (calyptobai) [reply server/bleep/src/webserver/answer.rs]: I've slightly simplified this in [3d772c3](https://github.com/BloopAI/bloop/pull/639/commits/3d772c307205938ca18ea42c5ad0f34ed4ce7490), we now rely on the parsed `SemanticQuery` stored on `Exchange`.

File: server/bleep/src/background.rs [hunk 28]
```diff
@@ -178,14 +225,40 @@ impl BoundSyncQueue {
...
+        self.1.queue.push(handle).await;
+        Ok(finished.recv_async().await?)
+    }
+
+    /// Index new branches specified in the `BranchFilter`.
+    ///
+    /// Unlike `sync_and_index`, this allows queueing multiple syncs
+    /// for the same repo multiple times to allow incrementally
+    /// syncing individual branches without restrictions.
+    pub(crate) async fn add_branches_for_repo(self, reporef: RepoRef, new_branches: BranchFilter) {
```
Comment (calyptobai) [inline server/bleep/src/background.rs:251]: Can we remove this function? Cargo warns that it is unused.

**Review by calyptobai — CHANGES_REQUESTED** [review]

**Review by ggordonhall — APPROVED** [review]
